eclipse-theia/theia
# Move Git history view to scm-extra package

#### What it does
The SCM view currently supports different SCM types through VSCode APIs.  However the History view supports Git only.  This PR refactors the history view, splitting apart the Git specific code from the history view.  This allows a single history view to be shown that tracks the selected repository regardless of whether the repository is a Git repository or a repository from another source manager type such as Mercurial.

This PR moves GitHistoryWidget to ScmHistoryWidget in the SCM package.  GitNavigableListWidget and other dependencies have also been moved.

The VSCode does not have API to support history.  Therefore we define our own API.  I could have implemented this in a similar way to the AmendSupport interface, where the provider provides an implementation.  However I decided instead to create a child container for each SCM type, so one child container for Git, one for Mercurial etc.  The history support implementations are bound in these containers.  The containers containing the repository options are now child containers to these.  This approach gives flexibility when adding extra SCM types, or extra an extra interface, or an implementation of an interface for a particular SCM type.  So, for example, someone could add support for SVN but not implement the History interface.  Someone else could add an extension that implements the History Support interface for SVN and bind it in the SVN child container.
  
#### How to test
This PR should be tested to ensure that it works with Git repositories exactly as before.  The only differences one should notice is that the view label has been changed from 'SCM' and 'Git History' to 'Source Control' and 'Source History'.  The label change was preferred by our UX team but if there are reasons for using the same label as VSCode then I can change it back and we will override it in our application.

#### Review checklist

- [x ] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
